File: packages/client/ts/src/spellguard.ts
// SPDX-License-Identifier: Apache-2.0

import {
  decryptFromVerifier,
  encryptForVerifier,
} from '@spellguard/amp/client';
import { loadProfile } from '@spellguard/amp/profile';
import type { AgentCard } from '@spellguard/ctls';
import type { LanguageModel } from 'ai';
import { Hono } from 'hono';
import {
  configure,
  createAttestationState,
  discoverAndConfigure,
  getClientKeyPair,
  getConfig,
  getOrCreateChannel,
  getVerifierPublicKey,
  runWithAttestationState,
} from './attestation';
import type { AttestationState } from './attestation';
import { runWithHops } from './hop-context';
import { setIntentDetectFn, setIntentDetectionModel } from './intent';
import type {
  IntentDetectionModelOrFactory,
  SpellguardConfigMode,
  SpellguardOptions,
} from './types';

/**
 * Create a Spellguard instance that manages configuration, model lifecycle,
 * and Hono middleware for Verifier callbacks, agent card, and health checks.
 *
 * Call `.middleware()` to get the Hono sub-app to mount on your router.
 * Call `.getModel()` to access the initialized model in route handlers.
 */
export function createSpellguard<E extends object = object, M = unknown>(
  options: SpellguardOptions<E, M>,
): SpellguardInstance<E, M> {
  let resolvedModel: M | undefined;
  let initPromise: Promise<void> | null = null;
  let initStartedAt = 0;

  // Per-instance attestation state: each createSpellguard() call gets
  // its own channel / config / discoveryConfig bucket, so multiple
  // instances hosted in the same worker (e.g. the the demo fleet) don't
  // overwrite each other on init or during outbound sends.  The
  // middleware wraps each request in this state via ALS.
  const instanceState: AttestationState = createAttestationState();

  const INIT_STALE_MS = 30_000;

  const SKIP_INIT_PATHS = new Set([
    '/_spellguard/health',
    '/.well-known/agent.json',
  ]);

  function getModel(): M {
    if (options.model && resolvedModel === undefined) {
      throw new Error(
        '[Spellguard] Model not initialized. Ensure middleware() has handled at least one non-skip request.',
      );
    }
    return resolvedModel as M;
  }

  async function initialize(env: E): Promise<void> {
    const cfg = resolveConfig(options.config, env);

    // Resolve the active profile (original | slim). Under the original
    // profile the existing attestation / channel / send code paths
    // remain authoritative; this call just surfaces the active profile
    // in startup logs and gives slim-profile follow-up commits a hook
    // to subscribe to. Pulls SPELLGUARD_* keys defensively from env
    // since the Worker env shape is generic.
    const envRecord = env as Record<string, unknown>;
    const profileBundle = loadProfile({
      SPELLGUARD_PROFILE: stringEnv(envRecord.SPELLGUARD_PROFILE),
      SPELLGUARD_TRANSPORT: stringEnv(envRecord.SPELLGUARD_TRANSPORT),
      SPELLGUARD_DIRECTORY: stringEnv(envRecord.SPELLGUARD_DIRECTORY),
      SPELLGUARD_IDENTITY: stringEnv(envRecord.SPELLGUARD_IDENTITY),
      SPELLGUARD_SLIM_GATEWAY_URL: stringEnv(
        envRecord.SPELLGUARD_SLIM_GATEWAY_URL,
      ),
      SPELLGUARD_DIR_URL: stringEnv(envRecord.SPELLGUARD_DIR_URL),
      SPELLGUARD_IDENTITY_ISSUER_URL: stringEnv(
        envRecord.SPELLGUARD_IDENTITY_ISSUER_URL,
      ),
    });
    console.log(
      `[Spellguard] Profile: ${profileBundle.profile} (transport=${profileBundle.transport.name}, directory=${profileBundle.directory.name}, identity=${profileBundle.identity.name})`,
    );

    // Auto-fill agentCard.url from config.selfUrl when empty
    const agentCard: AgentCard = options.agentCard.url
      ? options.agentCard
      : { ...options.agentCard, url: cfg.selfUrl };

    if (cfg.type === 'managed') {
      await discoverAndConfigure({
        agentId: cfg.agentId,
        agentSecret: cfg.agentSecret,
        // The signing key has to flow through here, not just live on
        // the ManagedConfig — discoverAndConfigure → configure() is
        // the only path that populates the channel's signingPrivateKey,
        // which createChannel() then uses to sign registration
        // evidence.  Dropping it here makes the channel fall back to
        // codeHash-as-seed, and the Verifier rejects with "Invalid
        // evidence signature" whenever it has the agent's real
        // public_key (i.e. every managed deployment).
        signingPrivateKey: cfg.signingPrivateKey,
        managementUrl: cfg.managementUrl,
        selfUrl: cfg.selfUrl,
        codeHash: cfg.codeHash,
        agentCard,
        platformAttestation: cfg.platformAttestation,
      });
    } else {
      configure({
        agentId: cfg.agentId,
        verifierUrl: cfg.verifierUrl,
        selfUrl: cfg.selfUrl,
        codeHash: cfg.codeHash,
        expectedVerifierImageHash: cfg.expectedVerifierImageHash,
        agentSecret: cfg.agentSecret,
        agentCard,
      });
      // Eagerly register with Verifier so this agent is discoverable
      // by other agents (matches the managed path behavior).
      const PRE_REG_TIMEOUT_MS = 15_000;
      try {
        await Promise.race([
          getOrCreateChannel(),
          new Promise<never>((_, reject) =>
            setTimeout(
              () => reject(new Error('pre-registration timed out')),
              PRE_REG_TIMEOUT_MS,
            ),
          ),
        ]);
        console.log('[Spellguard] Pre-registered with Verifier for discovery');
      } catch (error) {
        console.warn(
          `[Spellguard] Direct-config pre-registration failed (will retry on first send): ${error}`,
        );
      }
    }

    // Resolve the main model
    if (options.model) {
      resolvedModel = resolveModel(options.model, env);
    }

    // Set intent detection model if provided
    const rawIntentModel = options.intentDetectionModel;
    if (rawIntentModel) {
      const resolved = resolveIntentModel(rawIntentModel, env);
      if (typeof resolved === 'function') {
        setIntentDetectFn(resolved as (prompt: string) => Promise<string[]>);
      } else {
        setIntentDetectionModel(resolved as LanguageModel);
      }
    }

    if (options.onInitialized) {
      await options.onInitialized(env);
    }

    // No agent-side SLIM awareness: the agent only knows about its
    // Verifier URL (which points at the gateway in slim mode). The
    // gateway-backed verifier owns the slimName ↔ callbackUrl mapping
    // — it derives the agent's slimName from the CTLS registration
    // that already happened above and pushes a registry-update SLIM
    // control message to the gateway. Nothing here for the agent to do.

    console.log('[Spellguard] Initialization complete');
  }

  function middleware(): Hono<{ Bindings: E }> {
    const app = new Hono<{ Bindings: E }>();

    // Lazy init middleware, wrapped in this instance's AttestationState
    // so configure()/getOrCreateChannel() called from onMessage or
    // nested generateText calls always see THIS agent's channel and
    // config — not whichever createSpellguard instance initialized
    // most recently in the same worker.
    app.use('*', async (c, next) => {
      if (SKIP_INIT_PATHS.has(c.req.path)) {
        return next();
      }

      await runWithAttestationState(instanceState, async () => {
        if (initPromise && Date.now() - initStartedAt > INIT_STALE_MS) {
          console.warn('[Spellguard] Clearing stale init promise, retrying');
          initPromise = null;
        }

        if (!initPromise) {
          initStartedAt = Date.now();
          initPromise = initialize(c.env).catch((err) => {
            initPromise = null;
            throw err;
          });
        }

        await initPromise;
        await next();
      });
    });

    // Verifier callback endpoint
    app.post('/_spellguard/receive', async (c) => {
      const channelToken = c.req.header('X-Spellguard-Channel-Token');
      if (!channelToken) {
        return c.json({ error: 'Missing channel token' }, 401);
      }

      let body: {
        message?: unknown;
        encryptedMessage?: string;
        senderId: string;
        messageId: string;
        timestamp: number;
      };

      try {
        body = await c.req.json();
      } catch {
        return c.json({ error: 'Invalid JSON body' }, 400);
      }

      const { senderId, messageId } = body;
      // Gateway-opaque delivery (option a): when the Verifier encrypted to this
      // agent's registered key, decrypt it with our private key. A legacy
      // delivery carries a plaintext `message` instead — use it as-is.
      const wasEncrypted = typeof body.encryptedMessage === 'string';
      let message: unknown;
      if (wasEncrypted) {
        const keyPair = getClientKeyPair();
        if (!keyPair) {
          console.error(
            '[Spellguard] Received an encrypted message but have no agent keypair to decrypt with',
          );
          return c.json({ error: 'Failed to decrypt message' }, 400);
        }
        try {
          message = JSON.parse(
            decryptFromVerifier(
              body.encryptedMessage as string,
              keyPair.privateKeyHex,
            ),
          );
        } catch (err) {
          console.error(
            `[Spellguard] Failed to decrypt delivered message: ${err}`,
          );
          return c.json({ error: 'Failed to decrypt message' }, 400);
        }
      } else {
        message = body.message;
      }
      if (!message || !senderId) {
        return c.json({ error: 'Missing required fields' }, 400);
      }

      console.log(
        `[Spellguard] Received message ${messageId} from ${senderId}`,
      );

      try {
        // Extract hops + correlation id stamped by the Verifier so
        // that any outbound sendToAgent call within this async
        // context carries them forward.  Both fields are stamped on
        // the inbound payload by the Verifier router (see
        // packages/verifier/src/proxy/router.ts) — hops to enforce
        // MAX_MESSAGE_HOPS, correlation id to keep audit_logs rows
        // for one logical conversation grouped under a single
        // session.
        const hops =
          typeof message === 'object' && message !== null
            ? Number((message as Record<string, unknown>)._spellguardHops) || 0
            : 0;
        const correlationId =
          typeof message === 'object' && message !== null
            ? typeof (message as Record<string, unknown>)
                ._spellguardCorrelationId === 'string'
              ? ((message as Record<string, unknown>)
                  ._spellguardCorrelationId as string)
              : undefined
            : undefined;

        const response = await runWithHops(
          hops,
          () =>
            options.onMessage({
              message,
              senderId,
              model: getModel(),
              env: c.env,
            }),
          correlationId,
          // Carry the immediate sender into the trace context so any nested
          // routing this handler triggers excludes back-routing to the
          // sender (2-node cycle prevention — keeps the graph a DAG).
          senderId,
        );
        // Return leg: if the inbound was encrypted (so the Verifier supports it
        // and registered its X25519 key with us), encrypt our reply back TO the
        // Verifier. Otherwise reply in plaintext (legacy).
        const verifierPubkey = getVerifierPublicKey();
        if (wasEncrypted && verifierPubkey) {
          return c.json({
            success: true,
            encryptedResponse: encryptForVerifier(
              JSON.stringify(response),
              verifierPubkey,
            ),
          });
        }
        return c.json({ success: true, response });
      } catch (error) {
        console.error(`[Spellguard] Error handling message: ${error}`);
        return c.json(
          {
            error: 'Failed to process message',
            details: error instanceof Error ? error.message : String(error),
          },
          500,
        );
      }
    });

    // A2A Agent Card discovery
    app.get('/.well-known/agent.json', (c) => {
      const globalConfig = getConfig();
      const baseCard =
        !options.agentCard.url && globalConfig?.agentCard
          ? globalConfig.agentCard
          : options.agentCard;

      const card: AgentCard = {
        ...baseCard,
        ...(baseCard.url
          ? {}
          : { url: resolveConfig(options.config, c.env).selfUrl }),
        authentication: { schemes: ['spellguard-verifier'] },
      };
      return c.json(card);
    });

    // Health check
    app.get('/_spellguard/health', (c) => {
      const globalConfig = getConfig();
      return c.json({
        status: 'ok',
        agentId:
          globalConfig?.agentId ?? resolveConfig(options.config, c.env).agentId,
      });
    });

    return app;
  }

  return { middleware, getModel };
}

// ─── Helpers ───────────────────────────────────────────────────────

function resolveConfig<E extends object>(
  config: SpellguardConfigMode | ((env: E) => SpellguardConfigMode),
  env: E,
): SpellguardConfigMode {
  return typeof config === 'function' ? config(env) : config;
}

function stringEnv(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function resolveModel<E extends object, M>(
  modelOrFactory: ((env: E) => M) | { model: M },
  env: E,
): M {
  return typeof modelOrFactory === 'function'
    ? modelOrFactory(env)
    : modelOrFactory.model;
}

function resolveIntentModel<E extends object>(
  modelOrFactory: IntentDetectionModelOrFactory<E>,
  env: E,
): unknown {
  return typeof modelOrFactory === 'function'
    ? modelOrFactory(env)
    : modelOrFactory.model;
}

// ─── Public types ──────────────────────────────────────────────────

export interface SpellguardInstance<E extends object, M> {
  /** Hono sub-app with lazy init, Verifier callback, agent card, and health. */
  middleware(): Hono<{ Bindings: E }>;
  /** Get the initialized model. Throws if init hasn't completed yet. */
  getModel(): M;
}

/**
 * Verify that a request came from the Verifier.
 * In a full implementation, this would verify cryptographic signatures.
 */
export function verifyVerifierRequest(channelToken: string): boolean {
  return !!channelToken && channelToken.length > 0;
}

File: packages/client/py/spellguard_client/spellguard.py
# SPDX-License-Identifier: Apache-2.0

"""
spellguard_client - Spellguard Instance & FastAPI Integration

``create_spellguard()`` returns a ``SpellguardInstance`` that manages
configuration, model lifecycle, and a FastAPI app for Verifier callbacks,
agent card serving, and health checks.

Usage from an agent developer's perspective::

    from spellguard_client import create_spellguard
    from spellguard_client.ai import generate_text

    spellguard = create_spellguard(
        agent_card={"name": "my-agent", "url": "", "skills": [...]},
        config=lambda: {"type": "direct", "agent_id": "my-agent", ...},
        model=lambda: AsyncOpenAI(api_key="..."),
        on_message=on_message,
    )

    app = spellguard.app()          # FastAPI app with Spellguard routes
    model = spellguard.model         # The initialised AsyncOpenAI client

    @app.post("/chat")
    async def chat(request: Request):
        result = await generate_text(
            model=spellguard.model,
            model_name="gpt-4o",
            system="You are helpful.",
            prompt=body["message"],
        )
        return {"response": result.text}
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, Awaitable, Callable

from fastapi import FastAPI, Request, Response
from fastapi.responses import JSONResponse

from spellguard_ctls.types import (
    AgentCard,
    AgentCardAuthentication,
    AgentCardCapabilities,
    AgentCardSkill,
)

from .ai import set_current_correlation_id, set_current_hops
from .attestation import configure, discover_and_configure, get_config, get_or_create_channel
from .intent import set_intent_detect_fn, set_intent_detection_model
from .types import (
    DirectConfig,
    ManagedConfig,
    MessageContext,
    SpellguardConfig,
    SpellguardConfigMode,
    SpellguardDiscoveryConfig,
    SpellguardOptions,
)

logger = logging.getLogger("spellguard")


# ===================================================================
# Dict → dataclass converters
# ===================================================================


def _to_agent_card(val: AgentCard | dict[str, Any]) -> AgentCard:
    """Accept either an ``AgentCard`` dataclass or a plain dict."""
    if isinstance(val, AgentCard):
        return val
    if not isinstance(val, dict):
        raise TypeError(f"agent_card: expected AgentCard or dict, got {type(val)}")

    skills = [
        AgentCardSkill(**s) if isinstance(s, dict) else s
        for s in val.get("skills", [])
    ]
    caps = val.get("capabilities")
    if isinstance(caps, dict):
        caps = AgentCardCapabilities(
            streaming=caps.get("streaming"),
            push_notifications=caps.get("pushNotifications"),
        )
    auth = val.get("authentication")
    if isinstance(auth, dict):
        auth = AgentCardAuthentication(schemes=auth.get("schemes", []))

    return AgentCard(
        name=val.get("name", ""),
        url=val.get("url", ""),
        skills=skills,
        description=val.get("description"),
        version=val.get("version"),
        capabilities=caps,
        authentication=auth,
    )


def _to_config_mode(val: Any) -> SpellguardConfigMode:
    """Accept either a config dataclass or a plain dict."""
    if isinstance(val, (ManagedConfig, DirectConfig)):
        return val
    if isinstance(val, dict):
        if val.get("type") == "managed":
            return ManagedConfig(
                type="managed",
                agent_id=val.get("agent_id", ""),
                management_url=val.get("management_url", ""),
                self_url=val.get("self_url", ""),
                code_hash=val.get("code_hash", ""),
                agent_secret=val.get("agent_secret"),
                platform_attestation=val.get("platform_attestation"),
            )
        return DirectConfig(
            type="direct",
            agent_id=val.get("agent_id", ""),
            verifier_url=val.get("verifier_url", "http://localhost:3000"),
            self_url=val.get("self_url", ""),
            code_hash=val.get("code_hash", ""),
            expected_verifier_image_hash=val.get(
                "expected_verifier_image_hash", "sha384:dev-placeholder"
            ),
            agent_secret=val.get("agent_secret"),
        )
    raise TypeError(f"config: expected ManagedConfig, DirectConfig, or dict, got {type(val)}")


# ===================================================================
# SpellguardInstance
# ===================================================================


class SpellguardInstance:
    """Manages Spellguard configuration, model lifecycle, and a FastAPI
    app with lazy init, Verifier callbacks, agent card, and health check.
    """

    def __init__(self, options: SpellguardOptions) -> None:
        self._options = options
        self._resolved_model: Any | None = None
        self._init_promise: asyncio.Task[None] | None = None
        self._init_started_at: float = 0
        self._init_lock = asyncio.Lock()
        self._fastapi_app: FastAPI | None = None

        self._INIT_STALE_S = 30.0
        self._SKIP_INIT_PATHS = {
            "/_spellguard/health",
            "/.well-known/agent.json",
            "/health",
        }

    # --- public properties ------------------------------------------

    @property
    def model(self) -> Any:
        """The initialised model / client (e.g. ``AsyncOpenAI``).

        Available after the first non-skip request triggers lazy init.
        """
        return self._resolved_model

    # keep the old accessor for backwards compat
    def get_model(self) -> Any:
        return self._resolved_model

    def app(self) -> FastAPI:
        """Return (or create) the FastAPI app.

        The app already includes Spellguard routes
        (``/_spellguard/receive``, ``/.well-known/agent.json``,
        ``/_spellguard/health``) and a lazy-init middleware.
        Agent developers add their own routes directly to this app.
        """
        if self._fastapi_app is not None:
            return self._fastapi_app

        fastapi_app = FastAPI()
        self._fastapi_app = fastapi_app

        @fastapi_app.middleware("http")
        async def _lazy_init_middleware(
            request: Request, call_next: Any
        ) -> Response:
            if request.url.path not in self._SKIP_INIT_PATHS:
                await self._ensure_initialized()
            return await call_next(request)

        @fastapi_app.post("/_spellguard/receive")
        async def _receive(request: Request) -> Response:
            channel_token = request.headers.get("x-spellguard-channel-token")
            if not channel_token:
                return JSONResponse(
                    {"error": "Missing channel token"}, status_code=401
                )

            try:
                body = await request.json()
            except Exception:
                return JSONResponse(
                    {"error": "Invalid JSON body"}, status_code=400
                )

            message = body.get("message")
            sender_id = body.get("senderId")
            message_id = body.get("messageId")

            if not message or not sender_id:
                return JSONResponse(
                    {"error": "Missing required fields"}, status_code=400
                )

            logger.info(
                "[Spellguard] Received message %s from %s",
                message_id,
                sender_id,
            )

            try:
                # Extract hops + correlation id stamped by the
                # Verifier so any outbound _send_to_agent call
                # within this async context carries them forward.
                # Both fields ride on the inbound payload from
                # the Verifier router (see verifier/proxy/router.ts
                # forwardToRecipient).  hops drives the
                # MAX_MESSAGE_HOPS guard; correlation id keeps the
                # whole conversation under one audit_logs.correlation_id.
                hops = 0
                correlation_id: str | None = None
                if isinstance(message, dict):
                    raw_hops = message.get("_spellguardHops", 0)
                    hops = raw_hops if isinstance(raw_hops, int) else 0
                    raw_corr = message.get("_spellguardCorrelationId")
                    if isinstance(raw_corr, str) and raw_corr:
                        correlation_id = raw_corr

                hop_token = set_current_hops(hops)
                corr_token = set_current_correlation_id(correlation_id)
                try:
                    ctx = MessageContext(
                        message=message,
                        sender_id=sender_id,
                        model=self._resolved_model,
                    )
                    result = await self._options.on_message(ctx)
                finally:
                    # Reset to previous values even if on_message raises
                    from .ai import _current_correlation_id, _current_hops

                    _current_hops.reset(hop_token)
                    _current_correlation_id.reset(corr_token)
                return JSONResponse({"success": True, "response": result})
            except Exception as error:
                logger.error(
                    "[Spellguard] Error handling message: %s", error
                )
                return JSONResponse(
                    {
                        "error": "Failed to process message",
                        "details": str(error),
                    },
                    status_code=500,
                )

        @fastapi_app.get("/.well-known/agent.json")
        async def _agent_card() -> Response:
            global_config = get_config()
            base_card = (
                global_config.agent_card
                if (
                    not self._options.agent_card.url
                    and global_config
                    and global_config.agent_card
                )
                else self._options.agent_card
            )

            card_url = base_card.url
            if not card_url:
                cfg = self._resolve_config()
                card_url = cfg.self_url

            card_dict: dict[str, Any] = {
                "name": base_card.name,
                "url": card_url,
                "skills": [
                    {"id": s.id, "name": s.name, "description": s.description}
                    for s in base_card.skills
                ],
                "authentication": {"schemes": ["spellguard-verifier"]},
            }
            if base_card.description:
                card_dict["description"] = base_card.description
            if base_card.version:
                card_dict["version"] = base_card.version
            if base_card.capabilities:
                caps: dict[str, Any] = {}
                if base_card.capabilities.streaming is not None:
                    caps["streaming"] = base_card.capabilities.streaming
                if base_card.capabilities.push_notifications is not None:
                    caps["pushNotifications"] = (
                        base_card.capabilities.push_notifications
                    )
                if caps:
                    card_dict["capabilities"] = caps

            return JSONResponse(card_dict)

        @fastapi_app.get("/_spellguard/health")
        async def _health() -> Response:
            global_config = get_config()
            agent_id = (
                global_config.agent_id
                if global_config
                else self._resolve_config().agent_id
            )
            return JSONResponse({"status": "ok", "agentId": agent_id})

        return fastapi_app

    # --- internal ---------------------------------------------------

    def _resolve_config(self) -> SpellguardConfigMode:
        cfg = self._options.config
        if callable(cfg):
            raw = cfg()
        else:
            raw = cfg
        return _to_config_mode(raw) if isinstance(raw, dict) else raw

    async def _ensure_initialized(self) -> None:
        async with self._init_lock:
            if (
                self._init_promise is not None
                and time.time() - self._init_started_at > self._INIT_STALE_S
            ):
                logger.warning(
                    "[Spellguard] Clearing stale init promise, retrying"
                )
                self._init_promise = None

            if self._init_promise is None:
                self._init_started_at = time.time()
                self._init_promise = asyncio.ensure_future(self._initialize())

        try:
            await self._init_promise
        except Exception:
            async with self._init_lock:
                self._init_promise = None
            raise

    async def _initialize(self) -> None:
        cfg = self._resolve_config()

        # Auto-fill agentCard.url from config.selfUrl when empty
        agent_card = self._options.agent_card
        if not agent_card.url:
            from dataclasses import replace

            agent_card = replace(agent_card, url=cfg.self_url)

        if isinstance(cfg, ManagedConfig):
            await discover_and_configure(
                SpellguardDiscoveryConfig(
                    agent_id=cfg.agent_id,
                    agent_secret=cfg.agent_secret,
                    management_url=cfg.management_url,
                    self_url=cfg.self_url,
                    code_hash=cfg.code_hash,
                    agent_card=agent_card,
                    platform_attestation=cfg.platform_attestation,
                )
            )
        else:
            configure(
                SpellguardConfig(
                    agent_id=cfg.agent_id,
                    verifier_url=cfg.verifier_url,
                    self_url=cfg.self_url,
                    code_hash=cfg.code_hash,
                    expected_verifier_image_hash=cfg.expected_verifier_image_hash,
                    agent_secret=cfg.agent_secret,
                    agent_card=agent_card,
                )
            )
            # Eagerly register with Verifier so this agent is discoverable
            # by other agents via /agents/resolve/:name (matches the
            # managed path which does this inside discover_and_configure).
            try:
                await asyncio.wait_for(get_or_create_channel(), timeout=15.0)
                logger.info(
                    "[Spellguard] Pre-registered with Verifier for discovery"
                )
            except Exception as error:
                logger.warning(
                    "[Spellguard] Pre-registration failed "
                    "(will retry on first send): %s",
                    error,
                )

        # Resolve the main model
        if self._options.model is not None:
            m = self._options.model
            if callable(m) and not isinstance(m, dict):
                self._resolved_model = m()
            elif isinstance(m, dict) and "model" in m:
                self._resolved_model = m["model"]
            else:
                self._resolved_model = m

        # Set intent detection model if provided
        raw_intent = self._options.intent_detection_model
        if raw_intent is not None:
            if callable(raw_intent) and not isinstance(raw_intent, dict):
                resolved = raw_intent()
            elif isinstance(raw_intent, dict) and "model" in raw_intent:
                resolved = raw_intent["model"]
            else:
                resolved = raw_intent

            if callable(resolved):
                set_intent_detect_fn(resolved)
            else:
                set_intent_detection_model(resolved)

        if self._options.on_initialized:
            await self._options.on_initialized()

        logger.info("[Spellguard] Initialization complete")


# ===================================================================
# Factory function
# ===================================================================


def create_spellguard(
    options: SpellguardOptions | None = None,
    *,
    agent_card: AgentCard | dict[str, Any] | None = None,
    config: Any | None = None,
    on_message: Callable[[MessageContext], Awaitable[Any]] | None = None,
    model: Any | None = None,
    intent_detection_model: Any | None = None,
    on_initialized: Callable[..., Any] | None = None,
) -> SpellguardInstance:
    """Create a Spellguard instance.

    Can be called with a ``SpellguardOptions`` dataclass **or** with
    keyword arguments (the developer-friendly form)::

        # Keyword form (preferred for agent code):
        sg = create_spellguard(
            agent_card={"name": "my-agent", "url": "", "skills": [...]},
            config=lambda: {"type": "direct", ...},
            model=lambda: AsyncOpenAI(...),
            on_message=handle,
        )

        # Dataclass form (for library / adapter code):
        sg = create_spellguard(SpellguardOptions(...))
    """
    if options is not None:
        return SpellguardInstance(options)

    if agent_card is None or config is None or on_message is None:
        raise TypeError(
            "create_spellguard() requires either a SpellguardOptions "
            "object or at minimum agent_card=, config=, and on_message= "
            "keyword arguments."
        )

    return SpellguardInstance(
        SpellguardOptions(
            agent_card=_to_agent_card(agent_card),
            config=config,
            on_message=on_message,
            model=model,
            intent_detection_model=intent_detection_model,
            on_initialized=on_initialized,
        )
    )


def verify_verifier_request(channel_token: str) -> bool:
    """Verify that a request came from the Verifier."""
    return bool(channel_token) and len(channel_token) > 0
